### PR TITLE
Completed task-2

### PR DESCRIPTION
**Pull request by mahammadaliorudjov**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import com.es.phoneshop.model.Product;
+import org.codehaus.plexus.util.StringUtils;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Locale;
+
+public class ProductByDescriptionComparator implements Comparator<Product> {
+    private final String query;
+    private final String regex = " ";
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java]: `private static final String REGEX = " ";`

File: src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    public ProductByDescriptionComparator(String query) {
+        this.query = query;
+    }
+
+    @Override
+    public int compare(Product product1, Product product2) {
+        if (StringUtils.isBlank(query))
+            return 0;
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java]: add {  } brackets to the if clause and do not use if without { } brackets

File: src/main/java/com/es/phoneshop/model/ArrayListProductDao.java [hunk 3]
```diff
@@ -46,12 +50,22 @@ public Product getProduct(Long id) {
...
         return readWriteLock.read(() -> products.stream()
                 .filter(this::isProductAvailable)
+                .filter(product -> productDescriptionMatchesQuery(product, query))
+                .sorted(new ProductByDescriptionComparator(query))
+                .sorted(new ProductByOrderComparator(sortField, sortOrder))
                 .collect(Collectors.toList()));
     }
 
+    private boolean productDescriptionMatchesQuery(Product product, String query) {
+        String regex = " ";
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/model/ArrayListProductDao.java]: the approach is not bad, but lets create `private static final String REGEX = " ";`

File: src/main/java/com/es/phoneshop/model/Product.java [hunk 4]
```diff
@@ -95,4 +107,12 @@ public String getImageUrl() {
...
     }
+
+    public List<ProductPriceHistory> getProductPriceHistoryList() {
+        return productPriceHistoryList;
+    }
+
+    public void setProductPriceHistoryList(List<ProductPriceHistory> productPriceHistoryList) {
+        this.productPriceHistoryList = productPriceHistoryList;
+    }
 }
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/model/Product.java]: add an empty line

File: src/main/java/com/es/phoneshop/model/ProductDao.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.es.phoneshop.model;
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/model/ProductDao.java]: it might be an option to move DAO classes to the separate 'dao' package, on the same level of hieracrchy as 'model' package

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 6]
```diff
@@ -21,7 +28,20 @@ public void init(ServletConfig config) throws ServletException {
...
-        request.setAttribute("products", productDao.findProducts());
-        request.getRequestDispatcher("/WEB-INF/pages/productList.jsp").forward(request, response);
+        String query = request.getParameter(QUERY);
+        String sortField = request.getParameter(SORT_FIELD);
+        String sortOrder = request.getParameter(SORT_ORDER);
+        if (isSortingEnabled(sortField, sortOrder)) {
+            request.setAttribute(PRODUCTS, productDao.findProducts(query, null, null));
+        } else {
+            request.setAttribute(PRODUCTS, productDao.findProducts(query,
+                    SortField.valueOf(sortField.toUpperCase()),
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: paste this logic on a 32 line (use java ternary operation, for example), same for sortOrder. It gonna help to avoid if else clause

File: src/main/java/com/es/phoneshop/web/ProductPriceHistoryPageServlet.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        long id = Long.parseLong(request.getPathInfo().substring(1));
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/web/ProductPriceHistoryPageServlet.java]: hardcoded 1

File: src/main/webapp/WEB-INF/pages/productDetails.jsp [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            <td>${product.stock}</td>
+        </tr>
+        <tr>
+            <td>Price</td>
+            <td class="price">
+                <fmt:formatNumber value="${product.price}" type="currency" currencySymbol="${product.currency.symbol}"/>
+            </td>
+        </tr>
+    </table>
+</tags:master>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/pages/productDetails.jsp]: add an empty line

File: src/main/webapp/WEB-INF/pages/productNotFound.jsp [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<%@ page contentType="text/html;charset=UTF-8" %>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<%@ taglib prefix="tags" tagdir="/WEB-INF/tags" %>
+
+<tags:master pageTitle="404">
+    <h1>404</h1>
+    <p>Product not found</p>
+</tags:master>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/pages/productNotFound.jsp]: add an empty line

File: src/main/webapp/WEB-INF/pages/productPriceHistory.jsp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+          ${priceHistory.date}
+        </td>
+        <td class="price">
+            <fmt:formatNumber value="${priceHistory.price}" type="currency"
+                              currencySymbol="${priceHistory.currency.symbol}"/>
+        </td>
+      </tr>
+    </c:forEach>
+  </table>
+</tags:master>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/pages/productPriceHistory.jsp]: add an empty line

File: src/main/webapp/WEB-INF/pages/serverError.jsp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<%@ page contentType="text/html;charset=UTF-8" %>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+<%@ taglib prefix="fmt" uri="http://java.sun.com/jsp/jstl/fmt" %>
+<%@ taglib prefix="tags" tagdir="/WEB-INF/tags" %>
+
+<tags:master pageTitle="500">
+    <h1>500</h1>
+    <p>Internal Server Error</p>
+</tags:master>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/pages/serverError.jsp]: add an empty line

File: src/main/webapp/WEB-INF/tags/sortLink.tag [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<%@ tag trimDirectiveWhitespaces="true" %>
+<%@ attribute name="sort" required="true" %>
+<%@ attribute name="order" required="true" %>
+
+<a href="?sort=${sort}&order=${order}&query=${param.query}"
+   style="${sort eq param.sort and order eq param.order}">
+    ${order}
+</a>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/tags/sortLink.tag]: add an empty line

File: src/main/webapp/WEB-INF/web.xml [hunk 13]
```diff
@@ -9,10 +9,48 @@
...
+  </context-param>
+
+  <error-page>
+    <location>/WEB-INF/pages/productNotFound.jsp</location>
+    <exception-type>com.es.phoneshop.model.ProductNotFoundException</exception-type>
+  </error-page>
+
+  <error-page>
+    <error-code>500</error-code>
+    <location>/WEB-INF/pages/error/serverError.jsp</location>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/web.xml]: 500 error is not mapped

File: src/main/webapp/WEB-INF/tags/sortLink.tag [hunk 14]
```diff
@@ -0,0 +1,8 @@
+<%@ tag trimDirectiveWhitespaces="true" %>
+<%@ attribute name="sort" required="true" %>
+<%@ attribute name="order" required="true" %>
+
+<a href="?sort=${sort}&order=${order}&query=${param.query}"
+   style="${sort eq param.sort and order eq param.order}">
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/tags/sortLink.tag]: chosen asc/desc option doesn't become bold

File: src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+        return Long.compare(product2AmountOfMatches, product1AmountOfMatches);
+    }
+
+    private long countMatches(String[] words, String productDescription) {
+        return Arrays.stream(words)
+                .filter(word -> productDescription.toLowerCase(Locale.ROOT)
+                        .contains(word.toLowerCase(Locale.ROOT)))
+                .count();
+    }
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java]: as i can see, sorting is not working as expected. Try 'Siemens C' and here the last 'Siemens' (1/2 words)  should be higher then HTC (1/4) and sony ericsson (1/3)

File: src/main/java/com/es/phoneshop/comparators/ProductByOrderComparator.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
+package com.es.phoneshop.comparators;
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/comparators/ProductByOrderComparator.java:1]: add a Unit test class for this new class

File: src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
+package com.es.phoneshop.comparators;
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/comparators/ProductByDescriptionComparator.java:1]: add a Unit test class for this new class

File: src/test/java/com/es/phoneshop/model/ArrayListProductDaoTest.java [hunk 18]
```diff
@@ -199,16 +216,16 @@ public void testSaveExistingProductEmptyCode() {
...
 
-        assertFalse(productDao.findProducts().contains(productToDelete));
+        assertFalse(productDao.findProducts(QUERY, null, null).contains(productToDelete));
     }
 
     @Test
     public void testSaveSampleProductsInitializesCorrectly() {
-        assertEquals(INITIAL_PRODUCTS_COUNT, productDao.findProducts().size());
+        assertEquals(SAMSUNG_PRODUCTS_COUNT, productDao.findProducts(QUERY, null, null).size());
     }
```
Comment (YBylkou) [inline src/test/java/com/es/phoneshop/model/ArrayListProductDaoTest.java:280]: is it possible here to create a new test cases for the cases with fitering?

File: src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java [hunk 19]
```diff
@@ -57,9 +60,16 @@ public void testDoGetSetsProductsAttributeAndForwards() throws ServletException,
 
     @Test(expected = RuntimeException.class)
     public void testDoGetWhenDaoThrowsExceptionThrowsRuntimeException() throws ServletException, IOException {
+
+        when(request.getParameter(QUERY)).thenReturn(QUERY);
+        when(request.getParameter("sort")).thenReturn("description");
+        when(request.getParameter("order")).thenReturn("asc");
```
Comment (YBylkou) [inline src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java]: it is better to get rid of hardcoded values

File: src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java [hunk 20]
```diff
@@ -57,9 +60,16 @@ public void testDoGetSetsProductsAttributeAndForwards() throws ServletException,
...
         try (MockedStatic<ArrayListProductDao> mocked = mockStatic(ArrayListProductDao.class)) {
             mocked.when(ArrayListProductDao::getInstance).thenReturn(productDao);
-            when(productDao.findProducts()).thenThrow(new RuntimeException("Database error"));
+            when(productDao.findProducts(QUERY,
+                    SortField.valueOf("DESCRIPTION"),
+                    SortOrder.valueOf("ASC")))
```
Comment (YBylkou) [inline src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java]: it is better to get rid of hardcoded values

File: src/test/java/com/es/phoneshop/web/ProductPriceHistoryPageServletTest.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+    @Test(expected = ProductNotFoundException.class)
+    public void testDoGetNonExistingProductThrowsException() throws ServletException, IOException {
+        when(request.getPathInfo()).thenReturn(INVALID_PATH_INFO);
+        when(productDao.getProduct(NON_EXISTING_PRODUCT_ID)).thenReturn(product);
+        when(productDao.getProduct(NON_EXISTING_PRODUCT_ID).getProductPriceHistoryList())
+                .thenThrow(new ProductNotFoundException("Product not found"));
+
+        servlet.doGet(request, response);
+    }
+}
```
Comment (YBylkou) [inline src/test/java/com/es/phoneshop/web/ProductPriceHistoryPageServletTest.java]: add an empty line

**Review by YBylkou — CHANGES_REQUESTED** [review]

File: src/main/webapp/WEB-INF/web.xml [hunk 22]
```diff
@@ -9,10 +9,48 @@
...
+    <param-value>true</param-value>
+  </context-param>
+
+  <error-page>
+    <location>/WEB-INF/pages/productNotFound.jsp</location>
+    <exception-type>com.es.phoneshop.model.ProductNotFoundException</exception-type>
+  </error-page>
+
+  <error-page>
+    <error-code>500</error-code>
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/web.xml:53]: add also 404 mapping

**Review by YBylkou — CHANGES_REQUESTED** [review]

File: src/main/java/com/es/phoneshop/model/ProductNotFoundException.java [hunk 23]
```diff
@@ -1,4 +1,4 @@
-package com.es.phoneshop.model.product;
+package com.es.phoneshop.model;
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/model/ProductNotFoundException.java]: package exception

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 24]
```diff
@@ -21,7 +28,13 @@ public void init(ServletConfig config) throws ServletException {
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        request.setAttribute("products", productDao.findProducts());
-        request.getRequestDispatcher("/WEB-INF/pages/productList.jsp").forward(request, response);
+        String query = request.getParameter(QUERY);
+        SortField sortField = (request.getParameter(SORT_FIELD) != null)
+                ? SortField.valueOf(request.getParameter(SORT_FIELD).toUpperCase()) : null;
+        SortOrder sortOrder = (request.getParameter(SORT_ORDER) != null)
+                ? SortOrder.valueOf(request.getParameter(SORT_ORDER).toUpperCase()) : null;
```
Comment (YBylkou) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: move to generic method

File: src/test/java/com/es/phoneshop/comparators/ProductByDescriptionComparatorTest.java [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Test
+    public void compareWhenFirstProductHasMoreMatchesShouldReturnNegative() {
+        comparator = new ProductByDescriptionComparator(QUERY);
+        when(product1.getDescription()).thenReturn(APPLE_DESCRIPTION);
+        when(product2.getDescription()).thenReturn(SAMSUNG_DESCRIPTION);
+
+        assertTrue(comparator.compare(product1, product2) < ZERO);
+    }
```
Comment (YBylkou) [inline src/test/java/com/es/phoneshop/comparators/ProductByDescriptionComparatorTest.java:39]: add one more test case below zero

File: src/test/java/com/es/phoneshop/model/ArrayListProductDaoTest.java [hunk 26]
```diff
@@ -228,4 +295,13 @@ public void testSingletonConcurrentAccess() throws InterruptedException {
...
+
+    private boolean isSorted(List<Product> products, Comparator<Product> comparator) {
+        for (int i = 0; i < products.size() - 1; i++) {
+            if (comparator.compare(products.get(i), products.get(i + 1)) > 0) {
+                return false;
+            }
+        }
+        return true;
+    }
```
Comment (YBylkou) [inline src/test/java/com/es/phoneshop/model/ArrayListProductDaoTest.java]: Use iterator

File: src/main/webapp/WEB-INF/pages/productDetails.jsp [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+<jsp:useBean id="product" type="com.es.phoneshop.model.Product" scope="request"/>
+<tags:master pageTitle="Product List">
+    <p>
+        ${product.description}
+    </p>
+    <table>
+        <tr>
+            <td>Image</td>
+            <td>
+                <img class="product-tile" src="${product.imageUrl}">
```
Comment (YBylkou) [inline src/main/webapp/WEB-INF/pages/productDetails.jsp]: img is small

**Review by YBylkou — CHANGES_REQUESTED** [review]

**Review by YBylkou — APPROVED** [review]
